File: pigain/nodes/pig.py
#!/usr/bin/env python
import rospy
from visualization_msgs.msg import Marker
from visualization_msgs.msg import MarkerArray
from geometry_msgs.msg import PoseStamped

from pigain.msg import Node
from pigain.srv import Query, QueryResponse
from pigain.srv import BestNode, BestNodeResponse

from aeplanner.srv import Reevaluate

import numpy as np
from rtree import index

import gp

class PIGain:
    def __init__(self):
        self.s = rospy.Service('gp_query_server', Query, self.query_server)
        self.best_node_srv = rospy.Service('best_node_server', BestNode, self.best_node_srv_callback)
        self.reevaluate_client = rospy.ServiceProxy('reevaluate', Reevaluate)

        self.visualize_mean  = rospy.get_param('~visualize/mean',  False)
        self.visualize_sigma = rospy.get_param('~visualize/sigma', False)
        self.visualize_pts   = rospy.get_param('~visualize/pts',   False)
        self.resolution      = rospy.get_param('~visualize/resolution',   1)

        self.gain_sub = rospy.Subscriber('gain_node', Node, self.gain_callback)
        self.pose_sub = rospy.Subscriber('pose', PoseStamped, self.pose_callback)
        self.marker_pub = rospy.Publisher('pig_markers', MarkerArray, queue_size=10)
        self.mean_pub = rospy.Publisher('mean_markers', MarkerArray, queue_size=10)
        self.sigma_pub = rospy.Publisher('sigma_markers', MarkerArray, queue_size=10)

        if self.visualize_pts:
            rospy.Timer(rospy.Duration(1), self.rviz_callback)
        if self.visualize_mean or self.visualize_sigma:
            rospy.Timer(rospy.Duration(5), self.evaluate)

        # Get environment boundaries 
        try:
            self.min = rospy.get_param('boundary/min')
            self.max = rospy.get_param('boundary/max')
        except KeyError:
            rospy.logwarn("Boundary parameters not specified")
            rospy.logwarn("Defaulting to (-100, -100, 0), (100, 100, 3)...")
            self.min = [-100, -100, 0]
            self.max = [ 100,  100, 3]

        try:
            self.range = rospy.get_param('aep/gain/r_max') * 2
        except KeyError:
            rospy.logwarn("Range max parameter not specified")
            rospy.logwarn("Defaulting to 8 m...")

        self.bbx = (self.min[0], self.min[1], self.min[2], self.max[0], self.max[1], self.max[2])

        self.x = None
        self.y = None
        self.z = None

        self.hyperparam = gp.HyperParam(l = 1, sigma_f = 1, sigma_n = 0.1)

        # Create r-tree
        p = index.Property()
        p.dimension = 3
        self.idx = index.Index(properties = p)
        self.id = 0

        rospy.Timer(rospy.Duration(5), self.reevaluate_timer_callback)

    """ Save current pose of agent """
    def pose_callback(self, msg):
        self.x = msg.pose.position.x
        self.y = msg.pose.position.y
        self.z = msg.pose.position.z

    """ Reevaluate gain in all cached nodes that are closer to agent than self.range """
    def reevaluate_timer_callback(self, event):
        rospy.loginfo("reevaluate start")
        if self.x is None or self.y is None or self.z is None:
            rospy.logwarn("No position received yet...")
            rospy.logwarn("Make sure that 'pose' has been correctly mapped and that it is being published")
            return

        bbx = (self.x-self.range, self.y-self.range, self.z-self.range, 
               self.x+self.range, self.y+self.range, self.z+self.range)

        hits = self.idx.intersection(bbx, objects=True)

        reevaluate_list = []
        reevaluate_position_list = []
        for item in hits:
            if(item.object.gain > 2):
                reevaluate_position_list.append(item.object.position)
                reevaluate_list.append(item)
        try:
            res = self.reevaluate_client(reevaluate_position_list)
        except rospy.ServiceException, e:
            rospy.logerr("Calling reevaluate service failed")
            return

        for i, item in enumerate(reevaluate_list):
            item.object.gain = res.gain[i]
            item.object.yaw = res.yaw[i]

            self.idx.delete(item.id, (item.object.position.x, item.object.position.y, item.object.position.z))
            self.idx.insert(item.id, (item.object.position.x, item.object.position.y, item.object.position.z), obj=item.object)

        rospy.loginfo("reevaluate done")

    """ Insert node with estimated gain in rtree """
    def gain_callback(self, msg):
        self.idx.insert(self.id, (msg.position.x, msg.position.y, msg.position.z), obj=msg)
        self.id += 1

    """ Handle query to Gaussian Process """
    def query_server(self, req):
        bbx = (req.point.x-2, req.point.y-2, req.point.z-2, 
               req.point.x+2, req.point.y+2, req.point.z+2)
        y = np.empty((0))
        x = np.empty((0,3))
        hits = self.idx.intersection(bbx, objects=True)
        nearest = self.idx.nearest(bbx, 1, objects=True)

        for item in hits:
            y = np.append(y, [item.object.gain], axis=0)
            x = np.append(x, [[item.object.position.x, item.object.position.y, item.object.position.z]], axis = 0)

        yaw = 0
        for item in nearest:
            yaw = item.object.yaw

        if y.shape[0] == 0:
            response = QueryResponse()
            response.mu = 0
            response.sigma = 1
            return response

        xstar = np.array([[req.point.x, req.point.y, req.point.z]])

        mean, sigma = gp.gp(y, x, xstar, self.hyperparam, gp.sqexpkernel)

        response = QueryResponse()
        response.mu = mean
        response.sigma = sigma
        response.yaw = yaw

        return response

    """ Return all nodes with gain higher than req.threshold """
    def best_node_srv_callback(self, req):
        hits = self.idx.intersection(self.bbx, objects=True)

        best_gain = -1
        best_pose = None
        response = BestNodeResponse()
        for item in hits:
            if item.object.gain > req.threshold:
                response.best_node.append(item.object.position)
            if item.object.gain > best_gain:
                best_gain = item.object.gain

        response.gain = best_gain
        return response

    """ Evaluate potential information gain function over grid and publish it in rviz """
    def evaluate(self, event):
        y = np.empty((0))
        x = np.empty((0,3))
        xstar = np.empty((0,3))
        hits = self.idx.intersection(self.bbx, objects=True)
        for item in hits:
            y = np.append(y, [item.object.gain], axis=0)
            x = np.append(x, [[item.object.position.x, item.object.position.y, item.object.position.z]], axis = 0)

        xt = np.arange(self.min[0], self.max[0], self.resolution)
        yt = np.arange(self.min[1], self.max[1], self.resolution)
        zt = [1]

        for xx in xt:
            for yy in yt:
                for zz in zt:
                    xstar = np.append(xstar, [[xx, yy, zz]], axis = 0)

        mean, sigma = gp.gp(y, x, xstar, self.hyperparam, gp.sqexpkernel)
        mean_markers = MarkerArray()
        sigma_markers = MarkerArray()
        for id, pts in enumerate(zip(xstar, mean, sigma)):
            mean_markers.markers.append(self.np_array_to_marker(id, pts[0], pts[1], max(1-pts[2], 0)))
            # sigma_markers.markers.append(self.np_array_to_marker(id, pts[0], pts[2] * 2))
        
        self.mean_pub.publish(mean_markers)
        self.sigma_pub.publish(sigma_markers)

    """ Publish all cached nodes in rviz """
    def rviz_callback(self, event):
        markers = MarkerArray()
        hits = self.idx.intersection(self.bbx, objects=True)
        for item in hits:
            markers.markers.append(self.node_to_marker(item.id, item.object))
        
        self.marker_pub.publish(markers)


    def np_array_to_marker(self, id, p, v=0, a=0):
        marker = Marker()
        marker.header.frame_id = "map"
        marker.type = marker.CUBE
        marker.action = marker.ADD
        marker.id = id
        marker.scale.x = self.resolution
        marker.scale.y = self.resolution
        marker.scale.z = 0.1
        marker.color.r = v / 72.0
        marker.color.g = 0 
        marker.color.b = 0.5
        marker.color.a = a
        marker.pose.orientation.w = 1.0
        marker.pose.position.x = p[0]
        marker.pose.position.y = p[1]
        marker.pose.position.z = p[2]
        marker.lifetime = rospy.Time(10)

        return marker


    def node_to_marker(self, id, node):
        marker = Marker()
        marker.header.frame_id = "map"
        marker.type = marker.SPHERE
        marker.action = marker.ADD
        marker.id = id
        marker.scale.x = 0.4
        marker.scale.y = 0.4
        marker.scale.z = 0.4
        marker.color.r = node.gain / 72.0
        marker.color.g = 0.0
        marker.color.b = 0.5
        marker.color.a = 1.0
        marker.pose.orientation.w = 1.0
        marker.pose.position.x = node.position.x
        marker.pose.position.y = node.position.y
        marker.pose.position.z = node.position.z
        marker.lifetime = rospy.Time(1.2)

        return marker




if __name__ == '__main__':
    rospy.init_node('pigain', anonymous=True)
    pigain = PIGain()
    rospy.spin()

File: pigain/src/pig.cpp
#include <pigain/gp.h>
#include <pigain/pig.h>
#include <ros/ros.h>
#include <fstream>

#include <aeplanner_msgs/Reevaluate.h>

namespace pig
{
PIG::PIG(ros::NodeHandle& nh)
  : nh_(nh)
  , gp_srv_(nh_.advertiseService("gp_query_server", &PIG::gpQueryCallback, this))
  , best_node_srv_(nh_.advertiseService("best_node_server", &PIG::bestNodeCallback, this))
  , reevaluate_client_(nh_.serviceClient<aeplanner_msgs::Reevaluate>("reevaluate"))
  , gain_sub_(nh_.subscribe("gain_node", 100, &PIG::gainCallback, this))
  , pose_sub_(nh_.subscribe("pose", 10, &PIG::poseCallback, this))
  , marker_pub_(nh_.advertise<visualization_msgs::MarkerArray>("pig_markers", 10, true))
  , mean_pub_(nh_.advertise<visualization_msgs::MarkerArray>("mean_markers", 10, true))
  , sigma_pub_(nh_.advertise<visualization_msgs::MarkerArray>("sigma_markers", 10, true))
  , marker_pub_timer_(nh_.createTimer(ros::Duration(1), &PIG::rvizCallback, this))
  , mean_and_sigma_pub_timer_(
        nh.createTimer(ros::Duration(5), &PIG::evaluateCallback, this))
  , reevaluate_timer_(nh_.createTimer(ros::Duration(10), &PIG::reevaluateCallback, this))
  , pose_recieved_(false)
  , hyper_l_(1.0)
  , hyper_sigma_f_(1.0)
  , hyper_sigma_n_(0.1)
{
  // ---------- Get environment boundaries ----------
  resolution_ = nh_.param("visualize/resolution", 1.0);
  if (!nh_.getParam("aep/gain/r_max", range_))
  {
    ROS_WARN("Range max parameter not specified");
    range_ = 8;
    ROS_WARN("Defaulting to %f m...", range_);
  }
  range_ *= 2;  // Because of reasons TODO: Explain

  std::vector<double> min;
  std::vector<double> max;
  if (!nh_.getParam("boundary/min", min) || min.size() != 3)
  {
    ROS_WARN("Min boundary parameter not specified or incorrect");
    min.push_back(-100);
    min.push_back(-100);
    min.push_back(0);
    ROS_WARN("Defaulting to (%f, %f, %f)", min[0], min[1], min[2]);
  }
  if (!nh_.getParam("boundary/max", max) || max.size() != 3)
  {
    ROS_WARN("Max boundary parameter not specified or incorrect");
    max.push_back(100);
    max.push_back(100);
    max.push_back(3);
    ROS_WARN("Defaulting to (%f, %f, %f)", max[0], max[1], max[2]);
  }

  bbx_min_ = point(min[0], min[1], min[2]);
  bbx_max_ = point(max[0], max[1], max[2]);
}

bool PIG::gpQueryCallback(aeplanner_msgs::Query::Request& req,
                          aeplanner_msgs::Query::Response& res)
{

  box query_box(point(req.point.x - 2, req.point.y - 2, req.point.z - 2),
                point(req.point.x + 2, req.point.y + 2, req.point.z + 2));
  std::vector<value> hits;
  rtree_.query(boost::geometry::index::intersects(query_box), std::back_inserter(hits));
  std::vector<value> nearest;
  rtree_.query(boost::geometry::index::nearest(query_box, 1),
               std::back_inserter(nearest));

  Eigen::VectorXd y(hits.size());
  Eigen::Matrix<double, Eigen::Dynamic, 3> x(hits.size(), 3);

  for (size_t i = 0; i < hits.size(); ++i)
  {
    y(i) = std::get<1>(hits[i]).gain;
    geometry_msgs::Point position = std::get<1>(hits[i]).position;
    x(i, 0) = position.x;
    x(i, 1) = position.y;
    x(i, 2) = position.z;
  }

  double yaw = 0;
  for (value item : nearest)
  {
    yaw = std::get<1>(item).yaw;
  }

  if (y.size() == 0)
  {
    res.mu = 0;
    res.sigma = 1;
    return true;
  }

  Eigen::Matrix<double, Eigen::Dynamic, 3> x_star(1, 3);
  x_star(0, 0) = req.point.x;
  x_star(0, 1) = req.point.y;
  x_star(0, 2) = req.point.z;

  std::pair<double, double> gp_response =
      gp(y, x, x_star, hyper_l_, hyper_sigma_f_, hyper_sigma_n_);

  res.mu = gp_response.first;
  res.sigma = gp_response.second;
  res.yaw = yaw;

  return true;
}

bool PIG::bestNodeCallback(aeplanner_msgs::BestNode::Request& req,
                           aeplanner_msgs::BestNode::Response& res)
{
  box query_box(bbx_min_, bbx_max_);
  std::vector<value> hits;
  rtree_.query(boost::geometry::index::intersects(query_box), std::back_inserter(hits));

  double highest_gain = -1;
  for (value item : hits)
  {
    geometry_msgs::Point position = std::get<1>(item).position;
    double gain = std::get<1>(item).gain;
    if (gain > req.threshold)
    {
      res.best_node.push_back(position);
    }
    highest_gain = std::max(gain, highest_gain);
  }

  res.gain = highest_gain;
  return true;
}

void PIG::gainCallback(const aeplanner_msgs::Node::ConstPtr& msg)
{
  point p(msg->pose.pose.position.x, msg->pose.pose.position.y,
          msg->pose.pose.position.z);

  rtree_.insert(std::make_tuple(p, Node(*msg), id_++));
}

void PIG::poseCallback(const geometry_msgs::PoseStamped::ConstPtr& msg)
{
  pose_ = *msg;
  pose_recieved_ = true;
}

void PIG::evaluateCallback(const ros::TimerEvent& event)
{
  if (true || (mean_pub_.getNumSubscribers() == 0 && sigma_pub_.getNumSubscribers() == 0))
  {
    return;
  }

  box query_box(bbx_min_, bbx_max_);
  std::vector<value> hits;
  rtree_.query(boost::geometry::index::intersects(query_box), std::back_inserter(hits));

  Eigen::VectorXd y(hits.size());
  Eigen::Matrix<double, Eigen::Dynamic, 3> x(hits.size(), 3);

  for (size_t i = 0; i < hits.size(); ++i)
  {
    y(i) = std::get<1>(hits[i]).gain;
    geometry_msgs::Point position = std::get<1>(hits[i]).position;
    x(i, 0) = position.x;
    x(i, 1) = position.y;
    x(i, 2) = position.z;
  }

  Eigen::Matrix<double, Eigen::Dynamic, 3> x_star(
      int((bbx_max_.get<0>() - bbx_min_.get<0>()) / resolution_) +
          int((bbx_max_.get<1>() - bbx_min_.get<1>()) / resolution_) +
          int((bbx_max_.get<2>() - bbx_min_.get<2>()) / resolution_),
      3);
  size_t index = 0;
  for (double x = bbx_min_.get<0>(); x < bbx_max_.get<0>(); x += resolution_)
  {
    for (double y = bbx_min_.get<1>(); y < bbx_max_.get<1>(); y += resolution_)
    {
      for (double z = bbx_min_.get<2>(); z < bbx_max_.get<2>(); z += resolution_)
      {
        x_star(index, 0) = x;
        x_star(index, 1) = y;
        x_star(index, 2) = z;
        ++index;
      }
    }
  }

  std::pair<double, double> gp_response =
      gp(y, x, x_star, hyper_l_, hyper_sigma_f_, hyper_sigma_n_);

  visualization_msgs::MarkerArray mean_markers;
  visualization_msgs::MarkerArray sigma_markers;
  for (size_t i = 0; i < x_star.rows(); ++i)
  {
    mean_markers.markers.push_back(pointToMarker(
        i, x_star.row(i), gp_response.first, std::max(1.0 - gp_response.second, 0.0)));
    sigma_markers.markers.push_back(
        pointToMarker(i, x_star.row(i), gp_response.second * 2));
  }

  mean_pub_.publish(mean_markers);
  sigma_pub_.publish(sigma_markers);
}

void PIG::reevaluateCallback(const ros::TimerEvent& event)
{
  if (!pose_recieved_)
  {
    ROS_WARN("No position received yet...");
    ROS_WARN("Make sure that 'pose' has been correctly mapped and that it is being "
             "published");
    return;
  }

  geometry_msgs::Point position = pose_.pose.position;

  box query_box(point(position.x - range_, position.y - range_, position.z - range_),
                point(position.x + range_, position.y + range_, position.z + range_));
  std::vector<value> hits;
  rtree_.query(boost::geometry::index::intersects(query_box), std::back_inserter(hits));

  std::vector<value> reevaluate_list;
  std::vector<geometry_msgs::Point> reevaluate_position_list;
  for (value item : hits)
  {
    if (std::get<1>(item).gain > 0.1)
    {
      reevaluate_list.push_back(item);
      reevaluate_position_list.push_back(std::get<1>(item).position);
    }
  }

  aeplanner_msgs::Reevaluate srv;
  srv.request.point = reevaluate_position_list;
  if (!reevaluate_client_.call(srv))
  {
    ROS_ERROR("Calling reevaluate service failed");
    return;
  }

  for (size_t i = 0; i < reevaluate_list.size(); ++i)
  {
    value item = reevaluate_list[i];
    Node node(std::get<1>(item));
    node.gain = srv.response.gain[i];
    node.yaw = srv.response.yaw[i];

    rtree_.remove(item);
    rtree_.insert(std::make_tuple(std::get<0>(item), node, std::get<2>(item)));
  }
}

void PIG::rvizCallback(const ros::TimerEvent& event)
{
  if (marker_pub_.getNumSubscribers() == 0)
  {
    return;
  }

  visualization_msgs::MarkerArray markers;

  box query_box(bbx_min_, bbx_max_);
  std::vector<value> hits;
  boost::geometry::index::rtree<value, boost::geometry::index::rstar<16>> rtree = rtree_;
  rtree.query(boost::geometry::index::intersects(query_box), std::back_inserter(hits));

  for (value item : hits)
  {
    markers.markers.push_back(nodeToMarker(std::get<2>(item), std::get<1>(item)));
  }

  marker_pub_.publish(markers);
}

visualization_msgs::Marker PIG::pointToMarker(unsigned int id, Eigen::Vector3d point,
                                              double v, double a)
{
  visualization_msgs::Marker marker;
  marker.header.frame_id = "map";
  marker.type = marker.CUBE;
  marker.action = marker.ADD;
  marker.id = id;
  marker.scale.x = resolution_;
  marker.scale.y = resolution_;
  marker.scale.z = 0.1;
  marker.color.r = v;
  marker.color.g = 0;
  marker.color.b = 0.5;
  marker.color.a = a;
  marker.pose.orientation.w = 1.0;
  marker.pose.position.x = point[0];
  marker.pose.position.y = point[1];
  marker.pose.position.z = point[2];
  marker.lifetime = ros::Duration(10);

  return marker;
}

visualization_msgs::Marker PIG::nodeToMarker(unsigned int id, const Node node)
{
  visualization_msgs::Marker marker;
  marker.header.frame_id = "map";
  marker.type = marker.SPHERE;
  marker.action = marker.ADD;
  marker.id = id;
  marker.scale.x = 0.4;
  marker.scale.y = 0.4;
  marker.scale.z = 0.4;
  marker.color.r = node.gain / 32;
  marker.color.g = 0.0;
  marker.color.b = 0.5;
  marker.color.a = 1.0;
  marker.pose.orientation.w = 1.0;
  marker.pose.position.x = node.position.x;
  marker.pose.position.y = node.position.y;
  marker.pose.position.z = node.position.z;
  marker.lifetime = ros::Duration(3);

  return marker;
}
}  // namespace pig
